src/app/clusters/scenes/ExtensionFieldSetsImpl.cpp

```diff
@@ -0,0 +1,148 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "ExtensionFieldSetsImpl.h"
+
+namespace chip {
+namespace scenes {
+
+ExtensionFieldSetsImpl::ExtensionFieldSetsImpl() : ExtensionFieldSets() {}
+
+CHIP_ERROR ExtensionFieldSetsImpl::Serialize(TLV::TLVWriter & writer) const
+{
+    TLV::TLVType container;
+    ReturnErrorOnFailure(writer.StartContainer(TLV::ContextTag(kTagEFSArrayContainer), TLV::kTLVType_Structure, container));
+    ReturnErrorOnFailure(writer.Put(TLV::ContextTag(kTagEFSFieldNum), static_cast<uint8_t>(this->mFieldNum)));
+    if (!this->IsEmpty())
+    {
+        for (uint8_t i = 0; i < this->mFieldNum; i++)
+        {
+            if (!this->mEFS[i].IsEmpty())
+            {
+                ReturnErrorOnFailure(this->mEFS[i].Serialize(writer));
+            }
+        }
+    }
+
+    return writer.EndContainer(container);
+}
+
+CHIP_ERROR ExtensionFieldSetsImpl::Deserialize(TLV::TLVReader & reader)
+{
+    TLV::TLVType container;
+    ReturnErrorOnFailure(reader.Next(TLV::kTLVType_Structure, TLV::ContextTag(kTagEFSArrayContainer)));
+    ReturnErrorOnFailure(reader.EnterContainer(container));
+
+    ReturnErrorOnFailure(reader.Next(TLV::ContextTag(kTagEFSFieldNum)));
+    ReturnErrorOnFailure(reader.Get(this->mFieldNum));
+
+    if (!this->IsEmpty())
+    {
+        for (uint8_t i = 0; i < this->mFieldNum; i++)
+        {
+            ReturnErrorOnFailure(this->mEFS[i].Deserialize(reader));
```

Comment:
I added the a TestExentensionFieldSets.cpp locally and it will be pushed when I am done addressing all the comments. Indeed, an empty EFS[i] would make this fail. Because the array is always compressed on removal, no empty field should be found before this->FieldNum. If somehow an empty field is retrieved on deserializing, something went wrong.